hermes/src/store/types.rs

```diff
@@ -0,0 +1,303 @@
+use {
+    super::proof::wormhole_merkle::{
+        WormholeMerkleMessageProof,
+        WormholeMerkleProof,
+    },
+    anyhow::{
+        anyhow,
+        Result,
+    },
+    pythnet_sdk::{
+        accumulators::merkle::MerklePath,
+        hashers::keccak256_160::Keccak160,
+    },
+};
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum WormholePayload {
+    Merkle(WormholeMerkleProof),
+}
+
+impl WormholePayload {
+    pub fn try_from_bytes(bytes: &[u8], vaa_bytes: &[u8]) -> Result<Self> {
+        let magic = u32::from_be_bytes(bytes[0..4].try_into()?);
+        if magic != 0x41555756u32 {
+            return Err(anyhow!("Invalid magic"));
+        }
+
+        let message_type = u8::from_be_bytes(bytes[4..5].try_into()?);
+
+        if message_type != 0 {
+            return Err(anyhow!("Invalid message type"));
+        }
+
+        let state_index = u32::from_be_bytes(bytes[5..9].try_into()?);
+        let root_digest = bytes[9..29].try_into()?;
+
+        if bytes.len() > 29 {
+            return Err(anyhow!("Invalid message length"));
+        }
+
+        Ok(Self::Merkle(WormholeMerkleProof {
+            root: root_digest,
+            state_index,
+            vaa: vaa_bytes.to_vec(),
+        }))
+    }
+}
+
+pub type RawMessage = Vec<u8>;
+pub type MessageIdentifier = [u8; 32];
+
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub enum MessageType {
+    PriceFeed,
+    TwapPrice,
+}
+
+impl MessageType {
+    pub fn all() -> Vec<Self> {
+        // FIXME: This is a bit brittle, guard it in the future
+        vec![Self::PriceFeed, Self::TwapPrice]
+    }
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct WormholeMerkleState {
+    pub digest_proof: Vec<u8>,
+    pub tree:         Option<Vec<Vec<u8>>>,
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub struct MessageKey {
+    // -> this is the real message id
+    pub id:    MessageIdentifier, // -> this is price feed id
+    pub type_: MessageType,
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, PartialOrd, Ord)]
+pub struct MessageTime {
+    pub publish_time: UnixTimestamp,
+    pub slot:         Slot,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct ProofSet {
+    pub wormhole_merkle_proof: WormholeMerkleMessageProof,
+}
+
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct MessageState {
+    pub publish_time: UnixTimestamp,
+    pub slot:         Slot,
+    pub id:           MessageIdentifier,
+    pub type_:        MessageType,
+    pub message:      Message,
+    pub raw_message:  RawMessage,
+    pub proof_set:    ProofSet,
+}
+
+impl MessageState {
+    pub fn time(&self) -> MessageTime {
+        MessageTime {
+            publish_time: self.publish_time,
+            slot:         self.slot,
+        }
+    }
+
+    pub fn key(&self) -> MessageKey {
```

Comment:
this method doesn't belong on MessageState. This is an implementation detail of the storage system (and you could imagine other storage systems choosing a different key)

Reply:
I disagree, this is the key identifier of a message. Also MessageState is also part of the storage system interface. But let's discuss it later.